test/test_target.py

```diff
@@ -408,6 +408,52 @@ def test_pyopencl_execution_numpy_handling(ctx_factory):
     assert x[0] == 5.
 
 
+@pytest.mark.parametrize("target", [lp.PyOpenCLTarget, lp.ExecutableCTarget])
+def test_complex_support(ctx_factory, target):
+    knl = lp.make_kernel(
+            "{[i, i1, i2]: 0<=i,i1,i2<10}",
+            """
+            euler1[i]  = exp(3.14159265359j)
+            euler2[i] = (2.7182818284 ** (3.14159265359j))
+            euler1_real[i] = real(euler1[i])
+            euler1_imag[i] = imag(euler1[i])
+            real_times_complex[i] = in1[i]*(in2[i]*1j)
+            real_plus_complex[i] = in1[i] + (in2[i]*1j)
+            complex_div_complex[i] = (2jf + 7*in1[i])/(32jf + 37*in1[i])
+            complex_div_real[i] = (2jf + 7*in1[i])/in1[i]
+            real_div_complex[i] = in1[i]/(2jf + 7*in1[i])
+            tmp_sum[0] = sum(i1, 1.0*i1 + i1*1jf)*sum(i2, 1.0*i2 + i2*1jf)
+            """,
+            target=target())
+    knl = lp.set_options(knl, "return_dict")
+
+    n = 10
+
+    in1 = np.random.rand(n)
+    in2 = np.random.rand(n)
+
+    kwargs = {"in1": in1, "in2": in2}
+
+    if target == lp.PyOpenCLTarget:
+        knl = lp.set_options(knl, "write_cl")
+        evt, out = knl(cl.CommandQueue(ctx_factory()), **kwargs)
+    elif target == lp.ExecutableCTarget:
+        evt, out = knl(**kwargs)
+    else:
+        raise NotImplementedError("unsupported target")
+
+    np.testing.assert_allclose(out["euler1"], -1)
+    np.testing.assert_allclose(out["euler2"], -1)
+    np.testing.assert_allclose(out["euler1_real"], -1)
+    np.testing.assert_allclose(out["euler1_imag"], 0, atol=1e-10)
+    np.testing.assert_allclose(out["real_times_complex"], in1*(in2*1j))
+    np.testing.assert_allclose(out["real_plus_complex"], in1+(in2*1j))
+    np.testing.assert_allclose(out["complex_div_complex"], (2j+7*in1)/(32j+37*in1))
+    np.testing.assert_allclose(out["complex_div_real"], (2j + 7*in1)/in1)
+    np.testing.assert_allclose(out["real_div_complex"], in1/(2j + 7*in1))
+    np.testing.assert_allclose(out["tmp_sum"], (0.5*n*(n-1) + 0.5*n*(n-1)*1j) ** 2)
```

Comment:
There is also expression code gen testing based on randomly generated expressions [here](https://github.com/inducer/loopy/blob/95dafbfe69318b58c00fa2d30c2a07091efb06d4/test/test_expression.py#L247). I think it would be worthwhile (and not too fiddly) to hook up C99-complex to that test.

Reply:
That was handy, just ran those tests over the CTarget as well.